lib/screens/v2/candidate_generator/dup/departures.ex

```diff
@@ -501,37 +501,76 @@ defmodule Screens.V2.CandidateGenerator.Dup.Departures do
 
   defp get_overnight_schedules_for_section(_, _, _, _, _, _, _), do: []
 
+  defp get_overnight_departure_for_route(
+         _first_schedule_today,
+         nil,
+         _first_schedule_tomorrow,
+         route_id,
+         direction_id,
+         now
+       ) do
+    Logger.warn(
+      "[get_overnight_schedules_for_section] last_schedule_today not found. route_id=#{route_id} direction_id=#{direction_id} now=#{now}"
+    )
+
+    nil
+  end
+
+  defp get_overnight_departure_for_route(
+         first_schedule_today,
+         last_schedule_today,
+         first_schedule_tomorrow,
+         route_id,
+         direction_id,
+         now
+       ) do
+    cond do
+      is_nil(first_schedule_tomorrow) ->
+        nil
+
+      DateTime.compare(now, first_schedule_tomorrow.departure_time) == :gt ->
+        Logger.warn(
+          "[get_overnight_schedules_for_section] now is after first_schedule_tomorrow. route_id=#{route_id} direction_id=#{direction_id} now=#{now}"
+        )
+
+        nil
+
+      # Before 3am and between the `departure_time` for today's last schedule and tomorrow's first schedule
+      DateTime.compare(now, last_schedule_today.departure_time) == :gt and
+          DateTime.compare(now, first_schedule_tomorrow.departure_time) == :lt ->
+        %Departure{schedule: first_schedule_tomorrow}
+
+      # After 3am but before the first scheduled trip of the day.
+      DateTime.compare(now, first_schedule_today.departure_time) == :lt ->
+        %Departure{schedule: first_schedule_today}
+
+      true ->
+        nil
+    end
+  end
+
   defp get_today_tomorrow_schedules(
          fetch_params,
          fetch_schedules_fn,
-         tomorrow,
+         now,
          route_ids_serving_section
        ) do
     today =
-      case fetch_schedules_fn.(fetch_params, nil) do
+      case fetch_schedules_fn.(fetch_params, now) do
         {:ok, schedules} when schedules != [] ->
           # We want the last schedules of the current day.
           # Need to reverse the list of fetched schedules so that List.first/1 looks at the correct time of day.
```

Comment:
Since we're not reversing the list here anymore, I think we could remove this documentation